core/trino-main/src/main/java/io/trino/server/ui/OAuth2WebUiAuthenticationFilter.java

```diff
@@ -113,4 +120,22 @@ private void needAuthentication(ContainerRequestContext request)
         URI redirectLocation = service.startChallenge(request.getUriInfo().getBaseUri().resolve(CALLBACK_ENDPOINT));
         request.abortWith(Response.seeOther(redirectLocation).build());
     }
+
+    private boolean hasValidAudience(Claims claims)
+    {
+        if (validAudience.isEmpty()) {
+            return true;
+        }
+        Object claimsAudience = claims.get(Claims.AUDIENCE, Object.class);
+        if (claimsAudience == null) {
+            return false;
+        }
+        if (claimsAudience instanceof String) {
+            return claimsAudience.equals(validAudience.get());
+        }
+        if (claimsAudience instanceof List) {
+            return ((List<?>) claimsAudience).contains(validAudience.get());
+        }
+        return false;
```

Comment:
throw IllegalState?

Reply:
I would just return false.  We can't say it is for the auidence we expect, and the caller will fail with unauthorized anyway.